server/async_tcp.go

```diff
@@ -17,247 +16,211 @@ import (
 	"github.com/dicedb/dice/server/utils"
 )
 
-var cronFrequency time.Duration = 1 * time.Second
-var lastCronExecTime time.Time = utils.GetCurrentTime()
-
-const EngineStatusWAITING int32 = 1 << 1
-const EngineStatusBUSY int32 = 1 << 2
-const EngineStatusSHUTTINGDOWN int32 = 1 << 3
-const EngineStatusTRANSACTION int32 = 1 << 4
-
-var eStatus int32 = EngineStatusWAITING
-
-var connectedClients map[int]*core.Client
+type AsyncServer struct {
+	serverFD         int
+	ctx              context.Context
+	cancel           context.CancelFunc
+	wg               *sync.WaitGroup
+	maxClients       int
+	multiplexer      iomultiplexer.IOMultiplexer
+	connectedClients map[int]*core.Client
+	store            *core.Store
+	lastCronExecTime time.Time
+	cronFrequency    time.Duration
+}
 
-var asyncStore = core.NewStore()
+// NewAsyncServer initializes a new AsyncServer
+func NewAsyncServer(wg *sync.WaitGroup) *AsyncServer {
+	ctx, cancel := context.WithCancel(context.Background())
 
-func init() {
-	connectedClients = make(map[int]*core.Client)
+	return &AsyncServer{
+		ctx:              ctx,
+		cancel:           cancel,
+		wg:               wg,
+		maxClients:       20000,
+		connectedClients: make(map[int]*core.Client),
+		store:            core.NewStore(),
+		lastCronExecTime: utils.GetCurrentTime(),
+		cronFrequency:    1 * time.Second,
+	}
 }
 
-func setupUsers() {
-	var (
-		user *core.User
-		err  error
-	)
-	log.Info("setting up default user.", "password required", config.RequirePass != constants.EmptyStr)
-	if user, err = core.UserStore.Add(core.DefaultUserName); err != nil {
-		log.Fatal(err)
+// SetupUsers initializes the default user for the server
+func (s *AsyncServer) SetupUsers() error {
+	user, err := core.UserStore.Add(core.DefaultUserName)
+	if err != nil {
+		return err
 	}
-	if err = user.SetPassword(config.RequirePass); err != nil {
-		log.Fatal(err)
+	if err := user.SetPassword(config.RequirePass); err != nil {
+		return err
 	}
+	log.Info("default user set up", "password required", config.RequirePass != constants.EmptyStr)
+	return nil
 }
 
-// Waits on `core.WatchChannel` to receive updates about keys. Sends the update
-// to all the clients that are watching the key.
-// The message sent to the client will contain the new value and the operation
-// that was performed on the key.
-func WatchKeys(ctx context.Context, wg *sync.WaitGroup) {
-	defer wg.Done()
+// WatchKeys watches for changes in keys and notifies clients
+func (s *AsyncServer) WatchKeys() {
```

Comment:
```suggestion
func (s *AsyncServer) WatchKeys(ctx context.Context) {
```
Context can be passed here. And later we can think do we really want to wait on the context? I mean that case is required or not.